pmlc/dialect/tile/transforms/algebraic_opt.cc

```diff
@@ -0,0 +1,96 @@
+// Copyright 2020, Intel Corporation
+
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/DebugStringHelper.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "pmlc/dialect/tile/ir/ops.h"
+#include "pmlc/dialect/tile/transforms/pass_detail.h"
+#include "pmlc/dialect/tile/transforms/passes.h"
+#include "pmlc/util/logging.h"
+
+using namespace mlir; // NOLINT
+
+namespace pmlc::dialect::tile {
+
+namespace {
+
+// From:
+// %zero = tile.constant(0.0 : f64)
+// %0 = tile.contract add, mul, %zero, %I, %K
+// %1 = tile.add %0, %B
+// %2 = tile.relu %1
+// Into:
+// %0 = tile.contract add, mul, %B, %I, %K
+// %2 = tile.relu %0
+struct AddInitPattern final : public OpRewritePattern<AddOp> {
+  using OpRewritePattern<AddOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(AddOp op, PatternRewriter &rewriter) const {
+    return failure(
+        failed(matchAndRewritePermutation(op, rewriter, op.lhs(), op.rhs())) ||
```

Comment:
I think if lhs,rhs fails but rhs,lhs succeeds, we want the result to be success, this would make it a failure I believe (and actually not vheck rhs,lhs I think due to short circuit behavior).  I think you just need &&.

Reply:
I inverted this to check for success